Searching documents with `_doc_count` and synthetic source may lead to array_index_out_of_bounds_exception

### Elasticsearch Version

docker.elastic.co/elasticsearch/elasticsearch:8.7.0-eb092238-SNAPSHOT

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Linux (Docker)

### Problem Description

In some cases, after indexing a document with `_doc_count` into an index with synthetic source, an `array_index_out_of_bounds_exception` exception occurs when searching the index. This seems to be dependent on other documents indexed.

### Steps to Reproduce

Run the following API calls. The docs without `_doc_count` seem to matter; removing just one those stops the exception occurring.

```
DELETE /target

PUT /target
{
  "mappings": {
    "_source": {
      "mode": "synthetic"
    }
  }
}

POST /target/_doc
{
}

POST /target/_doc
{
}

POST /target/_doc
{
}

POST /target/_doc?refresh=true
{
  "_doc_count" : 1
}

GET /target/_search?error_trace
```

### Logs (if relevant)

```
# DELETE /target 200 OK
{
  "acknowledged": true
}
# PUT /target 200 OK
{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "target"
}
# POST /target/_doc 201 Created
{
  "_index": "target",
  "_id": "tlvLjYQBXlvXGQCu14Xw",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 0,
  "_primary_term": 1
}
# POST /target/_doc 201 Created
{
  "_index": "target",
  "_id": "t1vLjYQBXlvXGQCu2IUN",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 1,
  "_primary_term": 1
}
# POST /target/_doc 201 Created
{
  "_index": "target",
  "_id": "uFvLjYQBXlvXGQCu2IUk",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 2,
  "_primary_term": 1
}
# POST /target/_doc?refresh=true 201 Created
{
  "_index": "target",
  "_id": "uVvLjYQBXlvXGQCu2IU2",
  "_version": 1,
  "result": "created",
  "forced_refresh": true,
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 3,
  "_primary_term": 1
}
# GET /target/_search?error_trace 500 Internal Server Error
{
  "error": {
    "root_cause": [
      {
        "type": "array_index_out_of_bounds_exception",
        "reason": "Index 129 out of bounds for length 129",
        "stack_trace": """org.elasticsearch.ElasticsearchException$1: Index 129 out of bounds for length 129
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:640)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.SearchPhaseExecutionException.guessRootCauses(SearchPhaseExecutionException.java:150)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:622)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.ElasticsearchException.generateFailureXContent(ElasticsearchException.java:568)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.rest.RestResponse.build(RestResponse.java:175)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.rest.RestResponse.<init>(RestResponse.java:123)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.rest.RestResponse.<init>(RestResponse.java:102)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.rest.action.RestActionListener.onFailure(RestActionListener.java:55)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.rest.action.RestCancellableNodeClient$1.onFailure(RestCancellableNodeClient.java:96)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.client.internal.node.NodeClient$SafelyWrappedActionListener.onFailure(NodeClient.java:170)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:217)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:38)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$RunAfterActionListener.onFailure(ActionListener.java:358)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.raisePhaseFailure(AbstractSearchAsyncAction.java:751)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:728)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:418)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:760)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:512)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:349)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:48)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:642)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1642)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1372)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1504)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1479)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:48)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:917)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 129 out of bounds for length 129
	at org.apache.lucene.core@9.4.1/org.apache.lucene.codecs.lucene90.Lucene90PostingsReader$BlockDocsEnum.advance(Lucene90PostingsReader.java:545)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.DocCountFieldMapper$SyntheticFieldLoader.lambda$docValuesLoader$0(DocCountFieldMapper.java:158)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:638)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:121)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:242)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:190)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:130)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:70)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:147)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:76)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:672)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:644)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:495)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	... 6 more
"""
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "target",
        "node": "eX_04rcZTyWH1U0l5x-PvQ",
        "reason": {
          "type": "array_index_out_of_bounds_exception",
          "reason": "Index 129 out of bounds for length 129",
          "stack_trace": """java.lang.ArrayIndexOutOfBoundsException: Index 129 out of bounds for length 129
	at org.apache.lucene.core@9.4.1/org.apache.lucene.codecs.lucene90.Lucene90PostingsReader$BlockDocsEnum.advance(Lucene90PostingsReader.java:545)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.DocCountFieldMapper$SyntheticFieldLoader.lambda$docValuesLoader$0(DocCountFieldMapper.java:158)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:638)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:121)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:242)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:190)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:130)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:70)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:147)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:76)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:672)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:644)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:495)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:917)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
"""
        }
      }
    ],
    "caused_by": {
      "type": "array_index_out_of_bounds_exception",
      "reason": "Index 129 out of bounds for length 129",
      "caused_by": {
        "type": "array_index_out_of_bounds_exception",
        "reason": "Index 129 out of bounds for length 129",
        "stack_trace": """java.lang.ArrayIndexOutOfBoundsException: Index 129 out of bounds for length 129
	at org.apache.lucene.core@9.4.1/org.apache.lucene.codecs.lucene90.Lucene90PostingsReader$BlockDocsEnum.advance(Lucene90PostingsReader.java:545)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.DocCountFieldMapper$SyntheticFieldLoader.lambda$docValuesLoader$0(DocCountFieldMapper.java:158)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:638)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:121)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:242)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:190)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:130)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:70)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:147)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:76)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:672)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:644)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:495)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:917)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
"""
      },
      "stack_trace": """org.elasticsearch.ElasticsearchException$1: Index 129 out of bounds for length 129
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:640)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:416)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:760)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:512)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:349)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:48)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:642)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1642)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1372)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1504)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1479)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:48)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:917)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 129 out of bounds for length 129
	at org.apache.lucene.core@9.4.1/org.apache.lucene.codecs.lucene90.Lucene90PostingsReader$BlockDocsEnum.advance(Lucene90PostingsReader.java:545)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.DocCountFieldMapper$SyntheticFieldLoader.lambda$docValuesLoader$0(DocCountFieldMapper.java:158)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:638)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:121)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:242)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:190)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:130)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:70)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:147)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:76)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:672)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:644)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:495)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	... 6 more
"""
    },
    "stack_trace": """Failed to execute phase [query], all shards failed; shardFailures {[eX_04rcZTyWH1U0l5x-PvQ][target][0]: org.elasticsearch.transport.RemoteTransportException: [6e65998aa961][127.0.0.1:9300][indices:data/read/search[phase/query]]
Caused by: org.elasticsearch.search.fetch.FetchPhaseExecutionException: Fetch Failed [Error running fetch phase for doc [3]]
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:73)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:147)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:76)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:672)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:644)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:495)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:917)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 129 out of bounds for length 129
	at org.apache.lucene.core@9.4.1/org.apache.lucene.codecs.lucene90.Lucene90PostingsReader$BlockDocsEnum.advance(Lucene90PostingsReader.java:545)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.DocCountFieldMapper$SyntheticFieldLoader.lambda$docValuesLoader$0(DocCountFieldMapper.java:158)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:638)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:121)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:242)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:190)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:130)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:70)
	... 15 more
}
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:728)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:418)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:760)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:512)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:349)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:48)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:642)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1642)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1372)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1504)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1479)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:48)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:92)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:917)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: org.elasticsearch.ElasticsearchException$1: Index 129 out of bounds for length 129
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:640)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:416)
	... 20 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 129 out of bounds for length 129
	at org.apache.lucene.core@9.4.1/org.apache.lucene.codecs.lucene90.Lucene90PostingsReader$BlockDocsEnum.advance(Lucene90PostingsReader.java:545)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.DocCountFieldMapper$SyntheticFieldLoader.lambda$docValuesLoader$0(DocCountFieldMapper.java:158)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:638)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:121)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:242)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:190)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:130)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:70)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:147)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:76)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:672)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:644)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:495)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	... 6 more
"""
  },
  "status": 500
}
```

--- a/docs/changelog/91769.yaml
+++ b/docs/changelog/91769.yaml
@@ -1,0 +1,6 @@
+pr: 91769
+summary: Fix synthetic `_source` for sparse `_doc_count` field
+area: TSDB
+type: bug
+issues:
+ - 91731

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/get/100_synthetic_source.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/get/100_synthetic_source.yml
@@ -600,6 +600,7 @@ _doc_count:
             _source:
               mode: synthetic
 
+  # with _doc_count
   - do:
       index:
         index:   test
@@ -608,7 +609,6 @@ _doc_count:
         body:
           _doc_count: 3
           foo: bar
-
   - do:
       get:
         index: test
@@ -621,6 +621,98 @@ _doc_count:
       _source:
         _doc_count: 3
         foo: bar
+  - is_false: fields
+
+  # without _doc_count
+  - do:
+      index:
+        index:   test
+        id:      2
+        refresh: true
+        body:
+          foo: baz
+  - do:
+      get:
+        index: test
+        id:    2
+  - match: {_index: "test"}
+  - match: {_id: "2"}
+  - match: {_version: 1}
+  - match: {found: true}
+  - match:
+      _source:
+        foo: baz
+  - is_false: fields
+
+  # without immediately refreshing with _doc_count
+  - do:
+      index:
+        index:   test
+        id:      3
+        body:
+          _doc_count: 3
+          foo: qux
+  - do:
+      get:
+        index: test
+        id:    3
+  - match: {_index: "test"}
+  - match: {_id: "3"}
+  - match: {_version: 1}
+  - match: {found: true}
+  - match:
+      _source:
+        _doc_count: 3
+        foo: qux
+  - is_false: fields
+
+  # without immediately refreshing without _doc_count
+  - do:
+      index:
+        index:   test
+        id:      4
+        body:
+          foo: quux
+  - do:
+      get:
+        index: test
+        id:    4
+  - match: {_index: "test"}
+  - match: {_id: "4"}
+  - match: {_version: 1}
+  - match: {found: true}
+  - match:
+      _source:
+        foo: quux
+  - is_false: fields
+
+  # refresh all at once
+  - do:
+      indices.refresh: {}
+  - do:
+      get:
+        index: test
+        id:    3
+  - match: {_index: "test"}
+  - match: {_id: "3"}
+  - match: {_version: 1}
+  - match: {found: true}
+  - match:
+      _source:
+        _doc_count: 3
+        foo: qux
+  - is_false: fields
+  - do:
+      get:
+        index: test
+        id:    4
+  - match: {_index: "test"}
+  - match: {_id: "4"}
+  - match: {_version: 1}
+  - match: {found: true}
+  - match:
+      _source:
+        foo: quux
   - is_false: fields
 
 ---

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/400_synthetic_source.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/400_synthetic_source.yml
@@ -436,3 +436,63 @@ _source filtering:
   - match:
       hits.hits.0._source:
         kwd: foo
+
+---
+_doc_count:
+  - skip:
+      version: " - 8.5.99"
+      reason: introduced in 8.6.0
+
+  - do:
+      indices.create:
+        index: test
+        body:
+          settings:
+            number_of_replicas: 0
+          mappings:
+            _source:
+              mode: synthetic
+
+  - do:
+      index:
+        index:   test
+        id:      2
+        body:
+          foo: baz
+  - do:
+      index:
+        index:   test
+        id:      3
+        body:
+          foo: baz
+  - do:
+      index:
+        index:   test
+        id:      4
+        body:
+          foo: baz
+  - do:
+      index:
+        index:   test
+        id:      1
+        body:
+          _doc_count: 3
+          foo: bar
+  - do:
+      indices.refresh: {}
+
+  - do:
+      search:
+        index: test
+        body:
+          sort: foo.keyword
+  - is_false: hits.hits.0.fields
+  - is_false: hits.hits.1.fields
+  - match:
+      hits.hits.0._source:
+        _doc_count: 3
+        foo: bar
+  - match:
+      hits.hits.1._source:
+        foo: baz
+

--- a/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/400_synthetic_source.yml
+++ b/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/400_synthetic_source.yml
@@ -440,8 +440,8 @@ _source filtering:
 ---
 _doc_count:
   - skip:
-      version: " - 8.5.99"
-      reason: introduced in 8.6.0
+      version: " - 8.6.99"
+      reason: bug caused by many not having _doc_count fixed in 8.7.0
 
   - do:
       indices.create:

--- a/server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java
@@ -155,7 +155,15 @@ public class DocCountFieldMapper extends MetadataFieldMapper {
                 hasValue = false;
                 return null;
             }
-            return docId -> hasValue = docId == postings.advance(docId);
+            return docId -> {
+                if (docId < postings.docID()) {
+                    return hasValue = false;
+                }
+                if (docId == postings.docID()) {
+                    return hasValue = true;
+                }
+                return hasValue = docId == postings.advance(docId);
+            };
         }
 
         @Override

--- a/server/src/test/java/org/elasticsearch/index/mapper/DocCountFieldMapperTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/DocCountFieldMapperTests.java
@@ -110,4 +110,35 @@ public class DocCountFieldMapperTests extends MetadataMapperTestCase {
             }
         });
     }
+
+    public void testSyntheticSourceManyDoNotHave() throws IOException {
+        MapperService mapper = createMapperService(syntheticSourceMapping(b -> b.startObject("doc").field("type", "integer").endObject()));
+        List<Integer> counts = randomList(2, 10000, () -> randomBoolean() ? null : between(1, Integer.MAX_VALUE));
+        withLuceneIndex(mapper, iw -> {
+            int d = 0;
+            for (Integer c : counts) {
+                int doc = d++;
+                iw.addDocument(mapper.documentMapper().parse(source(b -> {
+                    b.field("doc", doc);
+                    if (c != null) {
+                        b.field(CONTENT_TYPE, c);
+                    }
+                })).rootDoc());
+            }
+        }, reader -> {
+            SourceLoader loader = mapper.mappingLookup().newSourceLoader();
+            assertTrue(loader.requiredStoredFields().isEmpty());
+            for (LeafReaderContext leaf : reader.leaves()) {
+                int[] docIds = IntStream.range(0, leaf.reader().maxDoc()).toArray();
+                SourceLoader.Leaf sourceLoaderLeaf = loader.leaf(leaf.reader(), docIds);
+                LeafStoredFieldLoader storedFieldLoader = StoredFieldLoader.empty().getLoader(leaf, docIds);
+                for (int docId : docIds) {
+                    String source = sourceLoaderLeaf.source(storedFieldLoader, docId).internalSourceRef().utf8ToString();
+                    int doc = (int) JsonXContent.jsonXContent.createParser(XContentParserConfiguration.EMPTY, source).map().get("doc");
+                    String docCountPart = counts.get(doc) == null ? "" : "\"_doc_count\":" + counts.get(doc) + ",";
+                    assertThat("doc " + docId, source, equalTo("{" + docCountPart + "\"doc\":" + doc + "}"));
+                }
+            }
+        });
+    }
 }

